infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,281 @@
+        synchronized(currEvent.getReplicationFsm()) {
+            while (true) {
+                if (canProcessEventForSession(session, currEvent.getEventId())) {
+                    break;
+                } else {
+                    try {
+                        currEvent.getReplicationFsm().wait();
+                    } catch (InterruptedException e) {
+                        log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                    }
+                }
+            }
+
+            LogReplicationState currState = currEvent.getReplicationFsm().getState();
+            if (currState.getType() == LogReplicationStateType.ERROR) {
+                log.info("Log Replication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            // TODO (Anny): consider strategy for continuously failing snapshot sync (never ending cancellation)
+
+            try {
+                LogReplicationState newState = currState.processEvent(currEvent);
+                log.trace("Transition from {} to {}", currState, newState);
+
+                currEvent.getReplicationFsm().transition(currState, newState);
+                currEvent.getReplicationFsm().setState(newState);
+                currEvent.getReplicationFsm().getNumTransitions().setValue(currEvent.getReplicationFsm().getNumTransitions().getValue() + 1);
+
+            } catch (org.corfudb.infrastructure.logreplication.replication.fsm.IllegalTransitionException illegalState) {
+                // Ignore LOG_ENTRY_SYNC_REPLICATED events for logging purposes as they will likely come in frequently,
+                // as it is used for update purposes but does not imply a transition.
+                if (!currEvent.getType().equals(LogReplicationEvent.LogReplicationEventType.LOG_ENTRY_SYNC_REPLICATED)) {
+                    log.error("Illegal log replication event {} when in state {}", currEvent.getType(), currState.getType());
+                }
+            }
+
+            // For testing purpose to notify the event generator the stop of the event.
+            if (currEvent.getType() == LogReplicationEvent.LogReplicationEventType.REPLICATION_STOP) {
+                synchronized (currEvent) {
+                    currEvent.notifyAll();
+                }
+            }
+
+            removeReplicationEventIdFromMap(session, currEvent.getEventId());
+            currEvent.getReplicationFsm().notifyAll();
+        }
+
+    }
+
+    private void removeReplicationEventIdFromMap(LogReplicationSession session, UUID currEventID) {
+        if (sessionToReplicationEventIdMap.get(session) != null && !sessionToReplicationEventIdMap.get(session).isEmpty() &&
+                sessionToReplicationEventIdMap.get(session).get(0).equals(currEventID)) {
+            sessionToReplicationEventIdMap.get(session).remove(0);
+        } else {
```

Comment:
No...its cleaning up the 2 Data Structures for maintaining order between the 0 delay and the non-zero delayed events..
This method gets called at the end of processReplicationTask(). So there is an event which needs to be cleaned up from the in-memory DS. Since the event themselves do not have any info if it had a non-zero delay or not, we just check one of the 2 DS for the replication event. If the event is found in one we delete it from there, else its bound to be in the other DS.